CMakeLists.txt

```
@@ -0,0 +1,135 @@
+#
+# CMakeLists file for OpenSSL
+# Copyright 2022-2023 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+#
+
+cmake_minimum_required(VERSION 3.14)
+
+project(OpenSSL LANGUAGES C CXX
+        DESCRIPTION "Open-source SSL and Cryptographic library"
+        HOMEPAGE_URL "https://github.com/openssl/openssl")
+
+# Configure the OpenSSL project
+if(OPENSSL_CONFIGURE_OPTIONS)
+  if (NOT PERL_EXECUTABLE) # Makes PERL_EXECUTABLE overridable
+    find_package(Perl 5.10.0 REQUIRED)
+  endif()
```

This logic skips validation of PERL_EXECUTABLE version if PERL_EXECUTABLE variable is defined.
Moreover, we just take it by belief - it can be literally any file.
Consider the following alternative:
```suggestion
    find_package(Perl 5.10.0 REQUIRED HINTS ${PERL_EXECUTABLE})
```

---

If it's been defined why would it be wrong? Is the user defining it telling a lie to cmake?

What's the consequence of that anyway? That it won't work, and it will tell you so?

---

Just `find_package(Perl 5.10.0 REQUIRED)`. It will use the cached `PERL_EXECUTABLE` if it is set.

---

> Just `find_package(Perl 5.10.0 REQUIRED)`. It will use the cached `PERL_EXECUTABLE` if it is set.

Yes, except the case where user configures the project with [CMAKE_FIND_USE_CMAKE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_USE_CMAKE_PATH.html#variable:CMAKE_FIND_USE_CMAKE_PATH)=False

---

> ~~~cmake
> find_package(Perl 5.10.0 REQUIRED PATHS "${PERL_EXECUTABLE}" NO_DEFAULT_PATH)

This is not valid CMake. I think you confuse `find_package` (which shall load the `FindPerl.cmake` module) and `find_program` (which (I assume) is used by the find module internally).

---

It is valid cmake. I was looking at this reference: https://cmake.org/cmake/help/latest/command/find_package.html#full-signature . I can be guilty of reading newest docs while you aim to support CMake v3.14, but it looks like even version 3.14 has full signature option for find_package. See https://cmake.org/cmake/help/v3.14/command/find_package.html#full-signature-and-config-mode .

---

> PERL_EXECUTABLE (or generally, find_program's cache variables) is always respected. It is the key input variable for supplying a particular executable and for avoiding searching again on repeated runs.
> CMAKE_FIND_USE_CMAKE only changes how the search is performed when the cache variable isn't set. For find_program, it turns off the use of CMAKE_PREFIX_PATH, CMAKE_PROGRAM_PATH, CMAKE_APPBUNDLE_PATH (user-supplied) cache variables.

This statement is not true for newest CMake revisions. I will quote part of this reference:
https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure

> (...) Describes search order by points 1 to 9. Pt. 1 omitted.
>  2. Search paths specified in cmake-specific cache variables. These are intended to be used on the command line with a -DVAR=VALUE. The values are interpreted as semicolon-separated lists. This can be skipped if NO_CMAKE_PATH is passed or by setting the CMAKE_FIND_USE_CMAKE_PATH to FALSE

It precisely says that cache values are ignored in the search process if CMAKE_FIND_USE_CMAKE_PATH = FALSE.

---

> It is valid cmake. I was looking at this reference: https://cmake.org/cmake/help/latest/command/find_package.html#full-signature . I can be guilty of reading newest docs while you aim to support CMake v3.14, but it looks like even version 3.14 has full signature option for find_package. See https://cmake.org/cmake/help/v3.14/command/find_package.html#full-signature-and-config-mode .

Okay, it is valid. But then let us look at the fine print:
- "User code should generally look for packages using the above [basic signature](https://cmake.org/cmake/help/v3.14/command/find_package.html#basic-signature)."
  Why not follow the recommendation?
- "The complete Config mode command signature is ..."
  AFAIU the full signature is for *Config* mode and will not find the `FindPerl.cmake` *Module*.
  The parameters `PATH` and `NO_DEFAULT_PATH` have to be interpreted with regard to location CMake config files, not perl executables.

---

> It precisely says that cache values are ignored in the search process if CMAKE_FIND_USE_CMAKE_PATH = FALSE.

It doesn't say that the target cache variable is ignored. In fact, your citations ends just before the color which explicitly lists the cache variables which are meant to be ignored. Full picture:

> Search paths specified in cmake-specific cache variables. These are intended to be used on the command line with a [-DVAR=VALUE](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-D). The values are interpreted as [semicolon-separated lists](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-lists). This can be skipped if NO_CMAKE_PATH is passed or by setting the [CMAKE_FIND_USE_CMAKE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_USE_CMAKE_PATH.html#variable:CMAKE_FIND_USE_CMAKE_PATH) to FALSE:
>
> - [CMAKE_PREFIX_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_PREFIX_PATH.html#variable:CMAKE_PREFIX_PATH)
> - [CMAKE_FRAMEWORK_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_FRAMEWORK_PATH.html#variable:CMAKE_FRAMEWORK_PATH)
> - [CMAKE_APPBUNDLE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_APPBUNDLE_PATH.html#variable:CMAKE_APPBUNDLE_PATH)

---

> We do not use `find_program` here.

Not directly, indeed. And that's the reason why `PERL_EXECUTABLE` doesn't belong into the parameter list (of `find_package(Perl)`).

The `FindPerl.cmake` module will do `find_program` and happily reuse `PERL_EXECUTABLE` if set:
https://github.com/Kitware/CMake/blob/007ff6867707ac5fa73e9709acf3a5ca00582af8/Modules/FindPerl.cmake#L39-L42